File: pythoncli/causewaybay/rpc.py
"""A small blocking JSON-RPC 2.0 client covering the methods the wallet needs."""

from __future__ import annotations

import itertools
import json
import time
from typing import Any

import requests

from . import errors

_REQUEST_IDS = itertools.count(1)

DEFAULT_TIMEOUT = 30.0
RECEIPT_POLL_INTERVAL = 1.5


def to_quantity(value: int) -> str:
    """Encode an integer as a minimal ``0x``-prefixed quantity."""
    return hex(value)


def parse_quantity(value: Any, context: str = "value") -> int:
    """Decode a ``0x``-prefixed quantity."""
    if not isinstance(value, str):
        raise errors.rpc_error(f"{context} returned {value!r} instead of a quantity")
    body = value[2:] if value[:2].lower() == "0x" else value
    if body == "":
        return 0
    # `int(body, 16)` would accept "-1", "1_0" and " 1 ". A JSON-RPC quantity is
    # hex digits and nothing else, so anything looser is a malformed response
    # rather than a number — the Rust side rejects the same inputs.
    if not all(c in "0123456789abcdefABCDEF" for c in body):
        raise errors.rpc_error(f"{context} returned an unparsable quantity {value}")
    return int(body, 16)


def field_int(obj: Any, field: str) -> int | None:
    """Read a hex quantity out of a JSON object field, tolerating absent fields."""
    if not isinstance(obj, dict) or field not in obj:
        return None
    try:
        return parse_quantity(obj[field], field)
    except errors.WalletError:
        return None


class RpcClient:
    def __init__(self, url: str, timeout: float = DEFAULT_TIMEOUT) -> None:
        self.url = url
        self.timeout = timeout
        self._session = requests.Session()
        self._session.headers.update(
            {"content-type": "application/json", "user-agent": "causewaybay-wallet/0.1.0"}
        )

    def call(self, method: str, params: list[Any] | None = None) -> Any:
        """Issue one JSON-RPC call and unwrap its ``result``."""
        payload = {
            "jsonrpc": "2.0",
            "id": next(_REQUEST_IDS),
            "method": method,
            "params": params if params is not None else [],
        }
        try:
            response = self._session.post(self.url, json=payload, timeout=self.timeout)
        except requests.RequestException as exc:
            raise errors.rpc_error(f"{method} request to {self.url} failed: {exc}") from exc

        text = response.text
        if not response.ok:
            raise errors.rpc_error(f"{method} returned HTTP {response.status_code}: {text[:200]}")
        try:
            body = json.loads(text)
        except json.JSONDecodeError:
            raise errors.rpc_error(f"{method} returned a non-JSON response: {text[:200]}") from None

        if isinstance(body, dict) and body.get("error"):
            rpc_err = body["error"]
            message = (
                rpc_err.get("message", "unknown RPC error")
                if isinstance(rpc_err, dict)
                else str(rpc_err)
            )
            code = rpc_err.get("code", 0) if isinstance(rpc_err, dict) else 0
            # Surface the funding case distinctly; it is the most common failure.
            if "insufficient funds" in message.lower():
                raise errors.insufficient_funds(message)
            raise errors.rpc_error(f"{method} failed ({code}): {message}")

        if not isinstance(body, dict) or "result" not in body:
            raise errors.rpc_error(f"{method} response has no result field")
        return body["result"]

    # ------------------------------------------------------------- shortcuts

    def chain_id(self) -> int:
        return parse_quantity(self.call("eth_chainId"), "eth_chainId")

    def block_number(self) -> int:
        return parse_quantity(self.call("eth_blockNumber"), "eth_blockNumber")

    def get_balance(self, address: str) -> int:
        return parse_quantity(self.call("eth_getBalance", [address, "latest"]), "eth_getBalance")

    def get_transaction_count(self, address: str) -> int:
        """Uses the pending block so back-to-back sends do not reuse a nonce."""
        return parse_quantity(
            self.call("eth_getTransactionCount", [address, "pending"]),
            "eth_getTransactionCount",
        )

    def gas_price(self) -> int:
        return parse_quantity(self.call("eth_gasPrice"), "eth_gasPrice")

    def estimate_gas(self, sender: str, to: str, value: int, data: bytes = b"") -> int:
        call: dict[str, Any] = {"from": sender, "to": to, "value": to_quantity(value)}
        if data:
            call["data"] = "0x" + data.hex()
        return parse_quantity(self.call("eth_estimateGas", [call, "latest"]), "eth_estimateGas")

    def eth_call(self, to: str, data: bytes) -> bytes:
        """Read-only contract call; returns the raw ABI-encoded return data."""
        result = self.call("eth_call", [{"to": to, "data": "0x" + data.hex()}, "latest"])
        if not isinstance(result, str):
            raise errors.rpc_error("eth_call did not return hex data")
        try:
            return bytes.fromhex(result[2:] if result[:2].lower() == "0x" else result)
        except ValueError as exc:
            raise errors.rpc_error(f"eth_call returned malformed hex: {exc}") from exc

    def send_raw_transaction(self, raw: bytes) -> str:
        result = self.call("eth_sendRawTransaction", ["0x" + raw.hex()])
        if not isinstance(result, str):
            raise errors.rpc_error("eth_sendRawTransaction did not return a hash")
        return result

    def get_transaction_by_hash(self, tx_hash: str) -> dict[str, Any] | None:
        return self.call("eth_getTransactionByHash", [tx_hash]) or None

    def get_transaction_receipt(self, tx_hash: str) -> dict[str, Any] | None:
        return self.call("eth_getTransactionReceipt", [tx_hash]) or None

    def wait_for_receipt(self, tx_hash: str, timeout: float = 180.0) -> dict[str, Any] | None:
        """Poll for a receipt until it appears or the deadline passes.

        Transient poll failures are retried rather than propagated: by the time
        this is called the transaction has already been broadcast, so reporting
        a single 502 as failure invites the user to send it a second time. Only
        a failure that persists to the deadline is surfaced.
        """
        deadline = time.monotonic() + timeout
        last_error: errors.WalletError | None = None
        while True:
            try:
                receipt = self.get_transaction_receipt(tx_hash)
            except errors.WalletError as exc:
                last_error = exc
            else:
                if receipt:
                    return receipt
                last_error = None
            if time.monotonic() + RECEIPT_POLL_INTERVAL > deadline:
                if last_error is not None:
                    raise errors.rpc_error(
                        f"{tx_hash} was broadcast, but polling for its receipt kept "
                        f"failing: {last_error.message}"
                    )
                return None
            time.sleep(RECEIPT_POLL_INTERVAL)

File: rustcli/core/src/rpc.rs
//! An async JSON-RPC 2.0 client covering the EVM methods the wallet needs.
//!
//! It shares the one HTTP client every chain uses (see [`crate::chain::http`]),
//! so a four-chain `balance --all` reuses connections instead of standing up a
//! stack per chain.

use std::time::Duration;

use alloy_primitives::{Address, U256};
use serde_json::{json, Value};

use crate::chain::http;
use crate::error::{self, Result};

pub struct RpcClient {
    url: String,
}

impl RpcClient {
    pub fn new(url: impl Into<String>) -> Result<Self> {
        Ok(RpcClient { url: url.into() })
    }

    pub fn url(&self) -> &str {
        &self.url
    }

    /// Issue one JSON-RPC call and unwrap its `result`.
    pub async fn call(&self, method: &str, params: Value) -> Result<Value> {
        http::json_rpc(&self.url, method, params).await
    }

    // ------------------------------------------------------------- shortcuts

    pub async fn chain_id(&self) -> Result<u64> {
        let value = self.call("eth_chainId", json!([])).await?;
        parse_quantity_u64(&value, "eth_chainId")
    }

    pub async fn block_number(&self) -> Result<u64> {
        let value = self.call("eth_blockNumber", json!([])).await?;
        parse_quantity_u64(&value, "eth_blockNumber")
    }

    pub async fn get_balance(&self, address: Address) -> Result<U256> {
        let value = self
            .call("eth_getBalance", json!([checksum(address), "latest"]))
            .await?;
        parse_quantity_u256(&value, "eth_getBalance")
    }

    /// Uses the pending block so back-to-back sends do not reuse a nonce.
    pub async fn get_transaction_count(&self, address: Address) -> Result<u64> {
        let value = self
            .call(
                "eth_getTransactionCount",
                json!([checksum(address), "pending"]),
            )
            .await?;
        parse_quantity_u64(&value, "eth_getTransactionCount")
    }

    pub async fn gas_price(&self) -> Result<U256> {
        let value = self.call("eth_gasPrice", json!([])).await?;
        parse_quantity_u256(&value, "eth_gasPrice")
    }

    pub async fn estimate_gas(
        &self,
        from: Address,
        to: Address,
        value: U256,
        data: &[u8],
    ) -> Result<u64> {
        let mut call = json!({
            "from": checksum(from),
            "to": checksum(to),
            "value": to_quantity(value),
        });
        if !data.is_empty() {
            call["data"] = json!(format!("0x{}", hex::encode(data)));
        }
        let result = self
            .call("eth_estimateGas", json!([call, "latest"]))
            .await?;
        parse_quantity_u64(&result, "eth_estimateGas")
    }

    /// Read-only contract call; returns the raw ABI-encoded return data.
    pub async fn eth_call(&self, to: Address, data: &[u8]) -> Result<Vec<u8>> {
        let call = json!({"to": checksum(to), "data": format!("0x{}", hex::encode(data))});
        let result = self.call("eth_call", json!([call, "latest"])).await?;
        let text = result
            .as_str()
            .ok_or_else(|| error::rpc_error("eth_call did not return hex data"))?;
        hex::decode(text.trim_start_matches("0x"))
            .map_err(|e| error::rpc_error(format!("eth_call returned malformed hex: {e}")))
    }

    pub async fn send_raw_transaction(&self, raw: &[u8]) -> Result<String> {
        let payload = format!("0x{}", hex::encode(raw));
        let result = self
            .call("eth_sendRawTransaction", json!([payload]))
            .await?;
        result
            .as_str()
            .map(str::to_string)
            .ok_or_else(|| error::rpc_error("eth_sendRawTransaction did not return a hash"))
    }

    pub async fn get_transaction_by_hash(&self, hash: &str) -> Result<Option<Value>> {
        let result = self.call("eth_getTransactionByHash", json!([hash])).await?;
        Ok(if result.is_null() { None } else { Some(result) })
    }

    pub async fn get_transaction_receipt(&self, hash: &str) -> Result<Option<Value>> {
        let result = self
            .call("eth_getTransactionReceipt", json!([hash]))
            .await?;
        Ok(if result.is_null() { None } else { Some(result) })
    }

    /// Poll for a receipt until it appears or the deadline passes.
    ///
    /// Transient poll failures are retried rather than propagated: by the time
    /// this is called the transaction has already been broadcast, so reporting
    /// a single 502 as failure invites the user to send it a second time. Only
    /// a failure that persists to the deadline is surfaced.
    pub async fn wait_for_receipt(&self, hash: &str, timeout: Duration) -> Result<Option<Value>> {
        let deadline = std::time::Instant::now() + timeout;
        let interval = Duration::from_millis(1500);
        let mut last_error: Option<crate::error::Error>;
        loop {
            match self.get_transaction_receipt(hash).await {
                Ok(Some(receipt)) => return Ok(Some(receipt)),
                Ok(None) => last_error = None,
                Err(e) => last_error = Some(e),
            }
            if std::time::Instant::now() + interval > deadline {
                // Nothing but errors for the whole window: say why, and name the
                // hash so the transaction can still be looked up.
                return match last_error {
                    Some(e) => Err(error::rpc_error(format!(
                        "{} was broadcast, but polling for its receipt kept failing: {}",
                        hash, e.message
                    ))),
                    None => Ok(None),
                };
            }
            tokio::time::sleep(interval).await;
        }
    }
}

fn checksum(address: Address) -> String {
    address.to_checksum(None)
}

/// Encode a U256 as a minimal `0x`-prefixed quantity, per the JSON-RPC spec.
pub fn to_quantity(value: U256) -> String {
    format!("0x{value:x}")
}

pub fn parse_quantity_u256(value: &Value, context: &str) -> Result<U256> {
    let text = value.as_str().ok_or_else(|| {
        error::rpc_error(format!("{context} returned {value} instead of a quantity"))
    })?;
    let body = text.trim_start_matches("0x").trim_start_matches("0X");
    if body.is_empty() {
        return Ok(U256::ZERO);
    }
    U256::from_str_radix(body, 16)
        .map_err(|_| error::rpc_error(format!("{context} returned an unparsable quantity {text}")))
}

pub fn parse_quantity_u64(value: &Value, context: &str) -> Result<u64> {
    let big = parse_quantity_u256(value, context)?;
    u64::try_from(big)
        .map_err(|_| error::rpc_error(format!("{context} returned a value too large for u64")))
}

/// Read a hex quantity out of a JSON object field, tolerating absent fields.
pub fn field_u64(value: &Value, field: &str) -> Option<u64> {
    parse_quantity_u64(value.get(field)?, field).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encodes_quantities_minimally() {
        assert_eq!(to_quantity(U256::ZERO), "0x0");
        assert_eq!(to_quantity(U256::from(1)), "0x1");
        assert_eq!(to_quantity(U256::from(255)), "0xff");
        assert_eq!(to_quantity(U256::from(1024)), "0x400");
    }

    #[test]
    fn parses_quantities() {
        assert_eq!(parse_quantity_u256(&json!("0x0"), "t").unwrap(), U256::ZERO);
        assert_eq!(parse_quantity_u256(&json!("0x"), "t").unwrap(), U256::ZERO);
        assert_eq!(
            parse_quantity_u256(&json!("0x152"), "t").unwrap(),
            U256::from(338)
        );
        assert_eq!(parse_quantity_u64(&json!("0x19"), "t").unwrap(), 25);
        assert_eq!(
            parse_quantity_u256(&json!("0xde0b6b3a7640000"), "t").unwrap(),
            U256::from(1_000_000_000_000_000_000u64)
        );
    }

    #[test]
    fn rejects_junk_quantities() {
        assert!(parse_quantity_u256(&json!(42), "t").is_err());
        assert!(parse_quantity_u256(&json!(null), "t").is_err());
        assert!(parse_quantity_u256(&json!("0xnothex"), "t").is_err());
        // 2^64 does not fit in a u64 even though it is a valid U256.
        assert!(parse_quantity_u64(&json!("0x10000000000000000"), "t").is_err());
    }

    #[test]
    fn reads_optional_object_fields() {
        let receipt = json!({"blockNumber": "0x2a", "status": "0x1"});
        assert_eq!(field_u64(&receipt, "blockNumber"), Some(42));
        assert_eq!(field_u64(&receipt, "status"), Some(1));
        assert_eq!(field_u64(&receipt, "missing"), None);
    }

    #[test]
    fn client_keeps_its_url() {
        let client = RpcClient::new("http://localhost:8545").unwrap();
        assert_eq!(client.url(), "http://localhost:8545");
    }
}
